### PR TITLE
CXX-591: Added list_databases and list collections functionality

### PR DESCRIPTION
**Pull request by yhager**

### CODE REVIEW

File: src/mongocxx/client.cpp [hunk 1]
```diff
@@ -66,5 +66,11 @@ class database client::database(bsoncxx::stdx::string_view name) const & {
     return mongocxx::database(*this, name);
 }
 
+cursor client::list_databases() const {
+    bson_error_t error;
+    auto result = libmongoc::client_find_databases(_impl->client_t, &error);
```
Comment (amidvidy) [inline src/mongocxx/client.cpp]: if result is nullptr, we should throw an exception::operation containing the message and code. See collection::create_index for an example.

File: src/mongocxx/client.hpp [hunk 2]
```diff
@@ -167,6 +167,13 @@ class MONGOCXX_API client {
     MONGOCXX_INLINE class database operator[](bsoncxx::stdx::string_view name) const &;
     MONGOCXX_INLINE class database operator[](bsoncxx::stdx::string_view name) const && = delete;
 
+    ///
+    /// Enumerates the databases in the client
+    ///
+    /// @return mongocxx::cursor containing the databases information.
+    ///
+    cursor list_databases() const;
```
Comment (amidvidy) [inline src/mongocxx/client.hpp]: add ```@throws exception::operation if the underlying 'listIndexes' command fails```

File: src/mongocxx/client.hpp [hunk 3]
```diff
@@ -167,6 +167,13 @@ class MONGOCXX_API client {
     MONGOCXX_INLINE class database operator[](bsoncxx::stdx::string_view name) const &;
     MONGOCXX_INLINE class database operator[](bsoncxx::stdx::string_view name) const && = delete;
 
+    ///
+    /// Enumerates the databases in the client
```
Comment (amidvidy) [inline src/mongocxx/client.hpp]: period at the end of the sentence.

File: src/mongocxx/client.hpp [hunk 4]
```diff
@@ -167,6 +167,13 @@ class MONGOCXX_API client {
...
+    ///
+    /// @return mongocxx::cursor containing the databases information.
```
Comment (amidvidy) [inline src/mongocxx/client.hpp]: `A mongocxx::cursor containing a BSON document for each database. Each document contains Each document contains a name field with the database name, a sizeOnDisk field with the total size of the database file on disk in bytes, and an empty field specifying whether the database has any data.

`@see http://docs.mongodb.org/manual/reference/command/listDatabases/`

File: src/mongocxx/database.cpp [hunk 5]
```diff
@@ -44,6 +44,14 @@ void* database::implementation() const {
     return _impl->database_t;
 }
 
+cursor database::list_collections(bsoncxx::document::view filter) {
+    libbson::scoped_bson_t filter_bson{filter};
+    bson_error_t error;
+
+    auto result = libmongoc::database_find_collections(_impl->database_t,
```
Comment (amidvidy) [inline src/mongocxx/database.cpp]: throw exception::operation if result is nullptr, as above.

File: src/mongocxx/database.hpp [hunk 6]
```diff
@@ -113,9 +113,12 @@ class MONGOCXX_API database {
     ///
     /// Enumerates the collections in this database.
     ///
+    /// @see http://docs.mongodb.org/manual/reference/command/listCollections/
+    ///
+    /// @param filter an optional query expression to filter the returned list
```
Comment (amidvidy) [inline src/mongocxx/database.hpp]: Should be formatted like (also with minor wording change)

```
@param filter
   An optional query expression to filter the returned collections. 
```

File: src/mongocxx/database.hpp [hunk 7]
```diff
@@ -113,9 +113,12 @@ class MONGOCXX_API database {
...
     /// @return mongocxx::cursor containing the collection information.
```
Comment (amidvidy) [inline src/mongocxx/database.hpp]: Also add `@throws` similar to `list_databases`.

**Comment by amidvidy** [comment]
Hey @yhager, this is a good start. I have some comments regarding error handling and documentation formatting.

**Comment by acmorrow** [comment]
@amidvidy @yhager Please ensure that the new APIs are in conformance with the published specification. See https://github.com/mongodb/specifications/blob/master/source/enumerate-collections.rst

Should probably also do
https://github.com/mongodb/specifications/blob/master/source/enumerate-indexes.rst
?

**Comment by amidvidy** [comment]
Yeah, mongocxx::collection::list_indexes is also unimplemented, I will create a new ticket for that.

**Comment by amidvidy** [comment]
our list_collections interface is in conformance with the spec, as is our list_indexes method (though as I mentioned, it is unimplemented)

**Comment by yhager** [comment]
Updated based on the comments above.

**Comment by amidvidy** [comment]
merged! Great work @yhager. Thanks for your contributions.

**Pull request closed without merging**